tests/backend/test_states.py

```diff
@@ -282,6 +282,68 @@ def test_squeezed_coherent(self, setup_backend, hbar, batch_size, tol):
 
         assert np.allclose(res.flatten(), res_exact.flatten(), atol=tol, rtol=0)
 
+@pytest.mark.backends("fock", "gaussian")
+class TestNumberExpectation:
+    """Multimode photon-number expectation value tests"""
+
+    def test_number_expectation_vacuum(self, setup_backend, tol):
+        """Tests the expectation value of any photon number in vacuum is zero"""
+        backend = setup_backend(2)
+        state = backend.state()
+        assert np.allclose(state.number_expectation([0, 1]), 0.0, atol=tol, rtol=0)
+        assert np.allclose(state.number_expectation([0]), 0.0, atol=tol, rtol=0)
+        assert np.allclose(state.number_expectation([1]), 0.0, atol=tol, rtol=0)
+
+    def test_number_expectation_displaced_squeezed(self, setup_backend, tol):
+        """Tests the expectation value of photon numbers when there is no correlation"""
+        backend = setup_backend(2)
+        state = backend.state()
+        a0 = 0.3 + 0.1 * 1j
+        r0 = 0.2
+        phi0 = 0.6
+        a1 = 0.1 + 0.1 * 1j
+        r1 = 0.3
+        phi1 = 0.9
+        backend.prepare_displaced_squeezed_state(a0, r0, phi0, 0)
+        backend.prepare_displaced_squeezed_state(a1, r1, phi1, 1)
+        state = backend.state()
+        n0 = np.sinh(r0) ** 2 + np.abs(a0) ** 2
+        n1 = np.sinh(r1) ** 2 + np.abs(a1) ** 2
+        assert np.allclose(state.number_expectation([0, 1]), n0 * n1, atol=tol, rtol=0)
+        assert np.allclose(state.number_expectation([0]), n0, atol=tol, rtol=0)
+        assert np.allclose(state.number_expectation([1]), n1, atol=tol, rtol=0)
+
+    def test_number_expectation_two_mode_squeezed(self, setup_backend, tol):
```

Comment:
It might be worth adding 1-1 test cases for when there are:
* more than 2 modes in the state, `number_expectation` is called with **consecutive** modes
* more than 2 modes in the state, `number_expectation` **non-consecutive** modes

Was wondering if it was useful to also test having more than two modes in the `Fock` case.

Reply:
> Was wondering if it was useful to also test having more than two modes in the Fock case.

Would definitely be useful!

Reply:
Don't understand the first two bullet points in Antals suggestion.

Reply:
In particular why is it interesting to test consecutive or non consecutive modes.

Reply:
With both ones, the first thing that is interesting is that we have **more modes (>2)** in our system than we are requesting the expectation for. This basically tests that tracing out works really the way we expect it to work.

Then, **consecutive** modes are the basic test. We'd expect that the user usually just requests these.

However, there could potentially be something funky going on with  **non-consecutive** modes, that's why I was suggesting to include that as well. It's simply just a potential edge case that users could get into.